### PR TITLE
feat(claims): add runtime snapshot reconciliation

### PR DESCRIPTION
**Pull request by jonathanhaaswriter**
## Summary
- add replace-existing claim writes that retract omitted asserted claims for a runtime snapshot
- add source-event claim filtering through bootstrap and both SDKs so snapshot readback can stay scoped
- update the Jira posture helpers to write reconciled snapshots and read back only asserted claims from the current snapshot

## Validation
- python3 -m py_compile sdk/python/cerebro_sdk/client.py sdk/python/cerebro_sdk/jira.py sdk/python/examples/jira_posture_onboarding.py
- python smoke for Jira posture claim and finding helpers
- npx -y -p typescript tsc -p sdk/typescript/tsconfig.json
- typescript smoke for Jira posture claim and finding helpers
- make verify

### CODE REVIEW

**Comment by jonathanhaaswriter** [comment]
@droid Please perform a deep code review and security review of this PR. Leave inline comments only for high-confidence, actionable issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review this PR after the latest fixes. Focus on whether the prior code review and security findings are resolved, and call out any remaining high-confidence issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest pushed fixes. Focus on whether the prior findings on this PR are resolved and only report remaining high-confidence issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest pushed fixes. Focus on whether the prior findings on this PR are resolved and only report remaining high-confidence issues.

**Comment by jonathanhaaswriter** [comment]
### Droid sweep status

Latest sweep: `2026-04-29 22:09 UTC`.

- Unresolved non-outdated review threads: `0`
- CI `verify`: `SUCCESS`
- `droid-review`: `not present`

I’ll keep watching for new review findings/check failures and address them as they appear.

**Comment by jonathanhaaswriter** [comment]
Droid sweep audit complete (final pass)

- Re-scanned all live review threads, including outdated ones
- Verified each remaining thread against the current branch tip
- Resolved every actionable thread (or fixed it inline and pushed)
- 0 unresolved review threads across the whole open-PR stack at this time

PR #494 (fix/feedback-main-followup) and PR #495 (feat/droid-creation-workflows) received additional fixes for SSRF (DNS-rebind) and workflow branch-name validation in this pass.

**Comment by jonathanhaaswriter** [comment]
@droid review

**Comment by jonathanhaaswriter** [comment]
@droid please harden this PR end-to-end:

1. Re-review the diff for residual bugs, races, edge cases, security gaps, and untested paths.
2. Apply the smallest correct fixes for any high-confidence issues you find.
3. Strengthen test coverage: add regression tests for every prior review finding on this PR plus the new behavior, including failure paths, boundary inputs, concurrency, and not-found / typed-nil interface guards.
4. Run focused tests and `make verify`. If validators fail, fix until green.
5. Commit and push the fixes to this PR's branch (or, if you lack push permission in this workflow, leave concrete inline suggestions and a summary comment).

Do not change the PR scope, do not modify the default branch, and do not merge.

**Comment by jonathanhaaswriter** [comment]
@droid review

File: internal/statestore/postgres/projection.go [hunk 1]
```diff
@@ -134,6 +134,46 @@ DO UPDATE SET
...
+	if s == nil || s.db == nil {
+		return errors.New("postgres is not configured")
+	}
+	if err := s.ensureProjectionTables(ctx); err != nil {
+		return err
+	}
+	if _, err := s.db.ExecContext(ctx, `
+DELETE FROM entity_links
+WHERE from_urn = $1 AND relation = $2 AND to_urn = $3 AND tenant_id = $4 AND source_id = $5`,
+		fromURN, relation, toURN, tenantID, sourceID); err != nil {
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/statestore/postgres/projection.go]: Fixed in 0a7d5c5e8 by deleting Postgres projected links using the actual uniqueness key `(from_urn, relation, to_urn)` instead of tenant/source attributes.

File: internal/graphstore/kuzu/kuzu.go [hunk 2]
```diff
@@ -433,6 +433,51 @@ func (s *Store) UpsertProjectedLink(ctx context.Context, link *ports.ProjectedLi
...
+		return err
+	}
+	statement := fmt.Sprintf(
+		"MATCH (src:entity {urn: %s})-[r:relation {relation: %s, tenant_id: %s, source_id: %s}]->(dst:entity {urn: %s}) DELETE r",
+		cypherString(fromURN),
+		cypherString(relation),
+		cypherString(tenantID),
+		cypherString(sourceID),
+		cypherString(toURN),
+	)
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/graphstore/kuzu/kuzu.go:464]: Fixed in 0a7d5c5e8 by making Kuzu `DeleteProjectedLink` match only endpoints + relation; added an integration regression deleting after tenant/source changes.

**Comment by jonathanhaaswriter** [comment]
@droid review

Please re-review the runtime snapshot reconciliation deeply. Replace-existing semantics are notoriously easy to get subtly wrong:

1. **Retraction correctness**: verify that omitted asserted claims in a snapshot are reliably retracted for the runtime, that retractions cannot reach across runtimes/tenants, and that partial snapshots (writer crashes mid-call) cannot leave the store in an inconsistent state.
2. **Atomicity and concurrency**: confirm replace-existing writes are atomic at the persistence layer (single transaction or equivalent), that two concurrent snapshots for the same runtime cannot interleave, and that the projection mirror reflects the final consistent state.
3. **Source-event filtering**: re-check filtering through bootstrap and both SDKs so that snapshot read-back stays scoped to source events the caller is authorized for, and that filter parsing rejects ambiguous/empty inputs as `InvalidArgument`.
4. **Idempotency**: ensure repeated snapshots with identical input produce no churn (no spurious deletes/inserts, no fingerprint changes), and that fingerprints/IDs are deterministic across runs.
5. **Read-back correctness**: verify read-back returns only the asserted claims of the current snapshot, never resurrected/retracted prior claims, and remains stable under interleaved reads/writes.
6. **Error surface**: confirm the new endpoints classify errors correctly (`InvalidArgument` for malformed snapshots, NotFound for missing runtimes, retriable server errors for transient store failures) and don’t leak internal store details.
7. **Forward-compat with downstream PRs**: ensure the reconciliation contract aligns with PR 463 (writes), PR 464 (reads), PR 468/470 (graph helpers), and PR 471 (posture findings), so reconciliation can be safely composed in higher-level flows.

File: internal/claims/service.go
Comment (josh-writer) [inline internal/claims/service.go:1]: `replace_existing` reconciles by claim ID after upserting, corrupting relation projections.  
Generated relation claim IDs omit object_urn (runtime/type/subject/predicate only), so a snapshot changing assigned_to from Alice to Bob reuses the same claim ID: the new Bob link is upserted, the stored claim row is overwritten, and the old Alice link is never deleted. If callers provide different IDs for the same relation triple, the opposite can happen: the newly upserted link can be deleted when retracting the old claim because projections are keyed by (from, relation, to), not claim ID. Reconcile relation projections against relation keys and/or delete stale links before upserting new links.

**Review by josh-writer — APPROVED** [review]

**Review by josh-writer — CHANGES_REQUESTED** [review]
meant to request changes, not approve, from previous comment

**Comment by jonathanhaaswriter** [comment]
Closing this PR — its substantive feature work is already in `main` via the earlier consolidation rollups, and the unresolved factory-droid findings posted on this PR have been addressed in main via the follow-up PRs:

- #497 (`fix(github): drop non-cursor opaque from terminal audit checkpoints`)
- #499 (`fix: address post-merge feedback across pending stacks`)

PR #499 specifically addresses the findings posted on this stack:

- Reject malformed `root_urn` values in `GetEntityNeighborhood` (PR #460)
- Map `refuted` and `superseded` claim statuses to projection deletion (PR #463)
- Tenant-scope `ListClaims` reads (PR #464)
- Per-runtime serialization of `WriteClaims` (PR #474)
- Restore null-array normalization in `graph_summary` (PR #470)
- Surface blank-root errors from `graphLayering` (PR #468)

This branch has diverged ~50k lines from `main` and rebasing it cleanly is no longer practical; closing here keeps the open queue accurate. If anything from this stack still needs to land, please open a fresh PR cut from `main`.
